mne/decoding/search_light.py

```diff
@@ -0,0 +1,491 @@
+            its hyperplane: e.g. [estimators[ii].decision_function(X[..., ii])
+                                  for ii in range(n_estimators)]
+
+        Returns
+        -------
+        y_pred : array, shape (n_samples, n_estimators, n_slices, n_classes)
+            Predicted values for each estimator.
+
+        Notes
+        -----
+        This requires base_estimator to have a `decision_function` method.
+        """
+        return self._transform(X, 'decision_function')
+
+    def score(self, X, y):
+        """Returns the score obtained for each combination of estimators and
+        tested dimensions.
+
+        Parameters
+        ----------
+        X : array, shape (n_samples, n_features, n_estimators)
+            The input samples. For each data slice, the corresponding estimator
+            score the prediction:
+            e.g. [estimators[ii].score(X[..., ii], y)
+                  for ii in range(n_estimators)]
+        y : array, shape (n_samples,)
+            The target values.
+
+        Returns
+        -------
+        score : array, shape (n_samples, n_estimators, n_slices, n_score_dim)  # noqa
+            Score for each estimator / data slice couple.
+
+        Notes
+        -----
+        This requires base_estimator to have a `decision_function` method.
+        """
+        self._check_Xy(X)
+        # For predictions/transforms the parallelization is across the data and
+        # not across the estimators to avoid memory load.
+        parallel, p_func, n_jobs = parallel_func(_gl_score, self.n_jobs)
+        X_splits = np.array_split(X, n_jobs, axis=-1)
+        score = parallel(p_func(self.estimators_, x, y) for x in X_splits)
+
+        if n_jobs > 1:
+            score = np.concatenate(score, axis=1)
+        else:
+            score = score[0]
+        return score
+
+
+def _gl_transform(estimators, X, method):
+    """Transform the dataset by applying each estimator to all slices of
+    the data.
+
+    Parameters
+    ----------
+    X : array, shape (n_samples, n_features, n_estimators)
+        The training input samples. For each data slice, a clone estimator
+        is fitted independently.
+
+    Returns
+    -------
+    Xt : array, shape (n_samples, n_estimators)
+        Transformed values generated by each estimator.
+    """
+    n_sample, n_chan, n_iter = X.shape
+    for ii, est in enumerate(estimators):
+        # stack generalized data for faster prediction
+        X_stack = np.transpose(X, [1, 0, 2])
+        X_stack = np.reshape(X_stack, [n_chan, n_sample * n_iter]).T
+        transform = getattr(est, method)
+        _y_pred = transform(X_stack)
+        # unstack generalizations
+        if _y_pred.ndim == 2:
+            _y_pred = np.reshape(_y_pred, [n_sample, n_iter, _y_pred.shape[1]])
+        else:
+            shape = np.r_[n_sample, n_iter, _y_pred.shape[1:]].astype(int)
+            _y_pred = np.reshape(_y_pred, shape)
```

Comment:
these are our latest optimizations from december in the gat code right, at least conceptually?